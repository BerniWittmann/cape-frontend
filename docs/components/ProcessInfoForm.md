# ProcessInfoForm

A Form to show and edit basic information about a process

## Props

<!-- @vuese:ProcessInfoForm:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|process|The Process Object|`Object`|`true`|-|
|isNewProcess|Option whether this process needs to be created or updated|`Boolean`|`true`|-|

<!-- @vuese:ProcessInfoForm:props:end -->


## Methods

<!-- @vuese:ProcessInfoForm:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|submit|Submits the form and returns result on successful validation|callback with data on success or undefined if invalid|
|setFormPristine|Sets the Form Pristine|-|

<!-- @vuese:ProcessInfoForm:methods:end -->


