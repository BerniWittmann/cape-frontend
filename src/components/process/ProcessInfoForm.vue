<template>
  <el-card class="process-info">
    <h4>{{ $t('process.edit.info')}}</h4>
    <el-form ref="processForm" :model="process" label-position="top" :rules="rules">
      <el-form-item :label="$t('process.edit.name')" prop="name">
        <el-input v-model="process.name"></el-input>
      </el-form-item>
      <el-form-item :label="$t('process.edit.tags')" prop="tags">
        <tag v-for="tag in process.tags" :tag="tag" :key="tag.id" closable @close="removeTag(tag)"></tag>
        <el-select
                v-if="tagInputVisible"
                v-model="newTag"
                size="small"
                ref="tagSelect"
                @keyup.enter.native="addTag"
                @change="addTag"
                @blur="hideTagInput"
                filterable
        >
          <el-option
                  v-for="tag in availableTags"
                  :key="tag.id"
                  :label="tag.name"
                  :value="tag">
            <tag :tag="tag" size="mini"></tag>
          </el-option>
        </el-select>
        <el-button v-if="!tagInputVisible && availableTags.length > 0" class="button-new-tag" size="small" @click="showTagInput" icon="el-icon-plus">{{
          $t('process.edit.add_tag') }}
        </el-button>
      </el-form-item>
    </el-form>
  </el-card>
</template>

<script>
/* ============
 * Tag Settings Component
 * ============
 *
 * A component to edit the Tags
 */

import TagComponent from '@/components/Tag.vue'

export default {
  components: {
    Tag: TagComponent
  },

  props: {
    process: {
      type: Object,
      required: true
    }
  },

  data() {
    return {
      rules: {
        name: [
          { required: true, message: this.$t('process.edit.validation.name.required'), trigger: 'blur' }
        ]
      },
      newTag: undefined,
      tagInputVisible: false
    }
  },

  computed: {
    allTags() {
      return this.$store.state.tag.tags
    },

    usedTagIDs() {
      return this.process.tags.map((t) => t.id)
    },

    availableTags() {
      return this.allTags.filter(t => !this.usedTagIDs.includes(t.id))
    }
  },

  methods: {
    submit(cb) {
      this.$refs.processForm.validate((valid) => {
        if (valid) {
          return cb(this.process)
        } else {
          return cb()
        }
      })
    },

    setFormPristine() {
      this.$refs.processForm.resetFields()
    },

    removeTag(tag) {
      this.process.tags = this.process.tags.filter((t) => t.id !== tag.id)
    },

    addTag() {
      this.process.tags.push(this.newTag)
    },

    showTagInput() {
      this.tagInputVisible = true
      this.$nextTick(() => {
        this.$refs.tagSelect.focus()
      })
    },

    hideTagInput() {
      setTimeout(() => {
        this.tagInputVisible = false
        this.newTag = undefined
      }, 100)
    }
  }
}

</script>

<style scoped lang="scss">
.process-info {
  background-color: #EBEEF5;
  border-radius: 5px;
  padding: 30px 20px;
}

.button-new-tag {
  margin-left: 5px;
}
</style>
