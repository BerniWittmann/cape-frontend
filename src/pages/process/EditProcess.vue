<template>
  <v-layout>
    <h2 class="process-edit__title">
      <el-button type="text" icon="el-icon-arrow-left" @click="$router.back()">{{ $t('process.edit.back' )}}</el-button>
      <span v-if="isNewProcess">{{ $t('process.add.title')}}</span>
      <span v-else>{{ $t('process.edit.title', { name })}}</span>
    </h2>
    <el-row :gutter="20">
      <el-col :span="18">
        <process-modeler v-model="processData" @input="() => {}"></process-modeler>
      </el-col>
      <el-col :span="6">
        <process-info-form ref="processInfoForm" :process="process"></process-info-form>
      </el-col>
    </el-row>
    <el-row type="flex" justify="center">
      <el-col :span="6">
        <el-button type="success" @click.native="submit">{{ $t('process.edit.save') }}</el-button>
        <el-button v-if="!isNewProcess" @click.native="reset" type="danger" plain>{{ $t('process.edit.reset') }}
        </el-button>
      </el-col>
    </el-row>
  </v-layout>
</template>

<script>
/* ============
 * Edit Process Page
 * ============
 *
 * A page where a process can be edited
 */

import DefaultLayout from '@/layouts/Default.vue'
import ProcessInfoForm from '@/components/process/ProcessInfoForm.vue'
import ProcessModeler from '@/components/process/ProcessModeler.vue'

import ProcessService from '@/services/process'
import Process from '@/models/process'

export default {
  components: {
    VLayout: DefaultLayout,
    ProcessInfoForm,
    ProcessModeler
  },

  computed: {
    process() {
      if (this.isNewProcess) {
        return new Process()
      }
      return this.$store.state.process.activeProcess
    },
    isNewProcess() {
      return this.$route.name === 'process.new'
    }
  },

  data() {
    return {
      name: undefined,
      processData: {
        xml: undefined,
        svg: undefined
      }
    }
  },

  methods: {
    submit() {
      this.$refs.processInfoForm.submit((result) => {
        if (result) {
          const data = Process.create({
            ...this.process,
            ...this.processData
          })
          data.id = this.process.id
          data.name = result.name
          data.tags = result.tags
          if (!this.isNewProcess) {
            ProcessService.update(data).then(() => {
              this.reset()
            })
          } else {
            ProcessService.create(data).then(() => {
              this.$router.replace({
                name: 'process.edit',
                params: {
                  processID: this.$store.state.process.processes[this.$store.state.process.processes.length - 1].id
                }

              })
            })
          }
        }
      })
    },

    reset() {
      ProcessService.get(this.process).then(() => {
        this.name = this.process.name
        this.processData.xml = this.process.xml
        this.processData.svg = this.process.svg
        this.$refs.processInfoForm.setFormPristine()
      })
    }
  },

  beforeMount() {
    this.name = this.process.name
    this.processData.xml = this.process.xml
    this.processData.svg = this.process.svg
  }
}
</script>

<style scoped lang="scss">
.process-edit__title {
  display: flex;
  align-items: center;

  .el-button {
    margin-right: 3em;
  }
}
</style>
