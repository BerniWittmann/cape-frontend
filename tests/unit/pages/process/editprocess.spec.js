import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'
import moment from 'moment'
import EditProcess from '@/pages/process/EditProcess.vue'
import ProcessInfoForm from '@/components/process/ProcessInfoForm.vue'
import EmptySlotComponent from '../../EmptySlotComponent.vue'
import { Button } from 'element-ui'
import ProcessService from '@/services/process'

describe('Pages', () => {
  describe('EditProcess.vue', () => {
    let store
    let router
    let cmp

    const date = moment(moment().utc())

    beforeEach(() => {
      router = {
        push: jest.fn(),
        back: jest.fn()
      }
      store = {
        state: {
          process: {
            activeProcess: {
              id: '1',
              name: 'My Process',
              createdAt: date.clone().subtract(14, 'days'),
              lastEditedAt: date.clone().subtract(2, 'days'),
              tags: [{
                id: '42',
                name: 'First Tag',
                color: '#FF0000'
              }, {
                id: '43',
                name: 'Second Tag',
                color: '#FFFF00'
              }]
            }
          },
          tag: {
            tags: [{
              id: '42',
              name: 'First Tag',
              color: '#FF0000'
            }, {
              id: '43',
              name: 'Second Tag',
              color: '#FFFF00'
            }, {
              id: '44',
              name: 'Third Tag',
              color: '#0FFF00'
            }]
          }
        },
        dispatch: jest.fn().mockImplementation(() => ({
          then: (arg) => arg()
        }))
      }
      render()
    })

    function render() {
      cmp = shallowMount(EditProcess, {
        i18n,
        mocks: {
          $store: store,
          $router: router
        },
        stubs: {
          'v-layout': EmptySlotComponent,
          'el-tooltip': EmptySlotComponent,
          'el-button': Button
        }
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('shows the title', () => {
      const title = cmp.find('.process-edit__title')
      expect(title.exists()).toBeTruthy()
      expect(title.text()).toContain('process.edit.back process.edit.title')
    })
    it('has a back button', () => {
      const button = cmp.find('.process-edit__title button')
      expect(button.exists()).toBeTruthy()
      button.trigger('click')
      expect(router.back).toHaveBeenCalled()
    })
    it('renders the process info form', () => {
      expect(cmp.contains(ProcessInfoForm)).toBeTruthy()
    })
    it('has a save and a reset button', () => {
      const buttons = cmp.findAll('elrow-stub button')
      expect(buttons.length).toEqual(2)

      expect(buttons.at(0).text()).toContain('save')
      expect(buttons.at(1).text()).toContain('reset')
    })
    it('can reset the process', () => {
      cmp.vm.$refs.processInfoForm.setFormPristine = jest.fn()
      ProcessService.get = jest.fn().mockImplementation(() => ({
        then: (arg) => arg()
      }))

      const button = cmp.findAll('elrow-stub button').at(1)
      button.trigger('click')

      expect(ProcessService.get).toHaveBeenCalled()
      expect(cmp.vm.$refs.processInfoForm.setFormPristine).toHaveBeenCalled()
    })
    it('can save the process', () => {
      cmp.vm.$refs.processInfoForm.submit = jest.fn().mockImplementation((cb) => {
        cb(store.state.process.activeProcess)
      })
      cmp.vm.reset = jest.fn()
      ProcessService.update = jest.fn().mockImplementation(() => ({
        then: (arg) => arg()
      }))

      const button = cmp.findAll('elrow-stub button').at(0)
      button.trigger('click')

      expect(cmp.vm.$refs.processInfoForm.submit).toHaveBeenCalled()
      expect(ProcessService.update).toHaveBeenCalled()
      expect(cmp.vm.reset).toHaveBeenCalled()
    })
    it('does not save the process if validation failed', () => {
      cmp.vm.$refs.processInfoForm.submit = jest.fn().mockImplementation((cb) => {
        cb()
      })
      cmp.vm.reset = jest.fn()
      ProcessService.update = jest.fn().mockImplementation(() => ({
        then: (arg) => arg()
      }))

      const button = cmp.findAll('elrow-stub button').at(0)
      button.trigger('click')

      expect(cmp.vm.$refs.processInfoForm.submit).toHaveBeenCalled()
      expect(ProcessService.update).not.toHaveBeenCalled()
      expect(cmp.vm.reset).not.toHaveBeenCalled()
    })
    it('unsets the active process on destroy', () => {
      cmp.destroy()

      expect(store.dispatch).toHaveBeenCalledWith('process/unsetActive')
    })
  })
})
