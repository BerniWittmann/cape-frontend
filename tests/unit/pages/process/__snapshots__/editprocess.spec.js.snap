// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages EditProcess.vue renders 1`] = `
<div>
  <h2 class="process-edit__title"><button type="button" class="el-button el-button--text">
      <!----><i class="el-icon-arrow-left"></i><span>process.edit.back</span></button> process.edit.title</h2>
  <elrow-stub tag="div" gutter="20" justify="start" align="top">
    <elcol-stub span="18" tag="div">
      <p>TODO Modeler comes here </p>
    </elcol-stub>
    <elcol-stub span="6" tag="div">
      <processinfoform-stub process="[object Object]"></processinfoform-stub>
    </elcol-stub>
  </elrow-stub>
  <elrow-stub tag="div" type="flex" justify="center" align="top">
    <elcol-stub span="6" tag="div"><button type="button" class="el-button el-button--success">
        <!---->
        <!----><span>process.edit.save</span></button> <button type="button" class="el-button el-button--danger is-plain">
        <!---->
        <!----><span>process.edit.reset</span></button></elcol-stub>
  </elrow-stub>
</div>
`;
